rse-bash.Rmd

```diff
@@ -1256,801 +1176,895 @@ $ ls lengths.txt
-Note that we can't write this as:
+and then use an exclamation mark `!` followed by a number to repeat one of those commands:
 
 ```shell
-$ for filename in *.dat
-> do
->     $filename
->     head -n 100 $filename | tail -n 20
-> done
+$ !572
 ```
 
-because then the first time through the loop,
-when `$filename` expanded to `basilisk.dat`, the shell would try to run `basilisk.dat` as a program.
-Finally,
-the `head` and `tail` combination selects lines 81-100
-from whatever file is being processed
-(assuming the file has at least 100 lines).
+```shell
+wc -l *.csv | sort -n | head -n 3
+    8314 bellambi_temp.csv
+    8315 bellambi_prec.csv
+    8680 tuggeranong_temp.csv
+```
 
-> **Spaces in Names**
->
-> Spaces are used to separate the elements of the list
-> that we are going to loop over. If one of those elements
-> contains a space character, we need to surround it with
-> quotes, and do the same thing to our loop variable.
-> Suppose our data files are named:
->
-> ```text
-> red dragon.dat
-> purple unicorn.dat
-> ```
->
-> To loop over these files, we would need to add double quotes like so:
->
-> ```shell
-> $ for filename in "red dragon.dat" "purple unicorn.dat"
-> do
->     head -n 100 "$filename" | tail -n 20
-> done
-> ```
->
-> It is simpler just to avoid using spaces (or other special characters) in filenames.
->
-> The files above don't exist, so if we run the above code, the `head` command will be unable
-> to find them, however the error message returned will show the name of the files it is
-> expecting:
-> ```text
-> head: cannot open ‘red dragon.dat’ for reading: No such file or directory
-> head: cannot open ‘purple unicorn.dat’ for reading: No such file or directory
-> ```
->
-> Try removing the quotes around `$filename` in the loop above
-> to see the effect of the quote marks on spaces.
-> Note that we get a result from the loop command for unicorn.dat
-> when we run this code in the `creatures` directory:
->
-> ```text
-> head: cannot open ‘red’ for reading: No such file or directory
-> head: cannot open ‘dragon.dat’ for reading: No such file or directory
-> head: cannot open ‘purple’ for reading: No such file or directory
-> CGGTACCGAA
-> AAGGGTCGCG
-> CAAGTGTTCC
-> ```
+The shell prints the command it is going to run to standard error rather than to standard output,
+so that (for example) `!572 > results.txt` puts the command's output into a file
+*without* also writing the command.
+
+Having an accurate record of what we have done
+and a simple way to repeat it
+are yet another reason why people use the Unix shell.
+In fact,
+being able to repeat history is such a powerful idea
+that the shell gives us several ways to do it.
 
-We would like to modify each of the files in `data-shell/creatures`, but also save a version
-of the original files, naming the copies `original-basilisk.dat` and `original-unicorn.dat`.
-We can't use:
+-   `!head` re-runs the most recent command starting with `head`,
```

Comment:
This is where zsh is so nice because it will exapnd `!head` and show the full command =)